### PR TITLE
Fix a hash collision issue on the `const_field` query

### LINKED ISSUE #61530: ICE when using `feature(stdsimd)` and IFMA intrinsics on non-IFMA platforms
Hi, while working on https://github.com/dalek-cryptography/curve25519-dalek/pull/257 , I ran into an ICE.

On a Cannonlake CPU with `target_feature=+avx512ifma`, the branch in the linked PR works fine.  On a Skylake CPU (without IFMA), the feature-gating *should* disable all of the IFMA code at compile-time.  But instead, running `cargo test` causes an ICE:

```
~/c/curve25519-dalek (use_upstream_intrinsics|✔) $ cargo test --features "simd_backend"
   Compiling curve25519-dalek v1.1.4 (/home/hdevalence/code/curve25519-dalek)
warning: unused import: `ValidityCheck`
    --> src/ristretto.rs:1084:28
     |
1084 |     use traits::{Identity, ValidityCheck};
     |                            ^^^^^^^^^^^^^
     |
     = note: #[warn(unused_imports)] on by default

thread 'rustc' panicked at 'Forcing query with already existing DepNode.
- query-key: ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: All, def_id: None }, value: (Const { ty: [u32; 8], val: ByRef(AllocId(4713).0x0, Allocation { bytes: [0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0, 4, 0, 0, 0, 5, 0, 0, 0, 6, 0, 0, 0, 7, 0, 0, 0], relocations: Relocations(SortedMap { data: [] }), undef_mask: UndefMask { blocks: [4294967295], len: Size { raw: 32 } }, align: Align { pow2: 2 }, mutability: Immutable, extra: () }) }, field[0]) }
- dep-node: const_field(eb99858e4874f291-9a55d263e1a1abdf)', src/librustc/ty/query/plumbing.rs:548:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (6ffb8f53e 2019-06-03) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental -C target_cpu=skylake --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `curve25519-dalek`.
warning: build failed, waiting for other jobs to finish...
thread 'rustc' panicked at 'Forcing query with already existing DepNode.
- query-key: ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: All, def_id: None }, value: (Const { ty: [u32; 8], val: ByRef(AllocId(6310).0x0, Allocation { bytes: [0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 3, 0, 0, 0, 4, 0, 0, 0, 5, 0, 0, 0, 6, 0, 0, 0, 7, 0, 0, 0], relocations: Relocations(SortedMap { data: [] }), undef_mask: UndefMask { blocks: [4294967295], len: Size { raw: 32 } }, align: Align { pow2: 2 }, mutability: Immutable, extra: () }) }, field[0]) }
- dep-node: const_field(eb99858e4874f291-9a55d263e1a1abdf)', src/librustc/ty/query/plumbing.rs:548:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.37.0-nightly (6ffb8f53e 2019-06-03) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental -C target_cpu=skylake

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `curve25519-dalek`.

To learn more, run the command again with --verbose.
```

(I think the reason the error appears twice is just because the crate loads itself into its own `build.rs`, not anything else.)

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #61530

### CODE REVIEW

File: src/librustc/mir/interpret/value.rs [hunk 1]
```diff
@@ -43,14 +43,13 @@ pub enum ConstValue<'tcx> {
...
-    /// An allocation together with a pointer into the allocation.
-    /// Invariant: the pointer's `AllocId` resolves to the allocation.
+    /// An allocation together with an offset into the allocation.
     /// The alignment exists to allow `const_field` to have `ByRef` access to nonprimitive fields
     /// of `repr(packed)` structs. The alignment may be lower than the type of this constant.
     /// This permits reads with lower alignment than what the type would normally require.
     /// FIXME(RalfJ,oli-obk): The alignment checks are part of miri, but const eval doesn't really
     /// need them. Disabling them may be too hard though.
-    ByRef(Pointer, Align, &'tcx Allocation),
+    ByRef(Size, Align, &'tcx Allocation),
```
Comment (eddyb) [inline src/librustc/mir/interpret/value.rs]: Can you use a struct variant so you can name the `Size` field as `offset`?

File: src/test/incremental/simd_intrinsic_ice.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#![feature(repr_simd, platform_intrinsics)]
+
+// revisions:rpass1 rpass2
```
Comment (eddyb) [inline src/test/incremental/simd_intrinsic_ice.rs]: Maybe it'd make more sense to name this after the issue?

**Comment by eddyb** [comment]
r? @eddyb 

cc @pnkfelix This should fix #61594 as well.

**Review by eddyb — APPROVED** [review]
r=me with nits fixed (if at all)

File: src/test/incremental/simd_intrinsic_ice.rs (hunk 2 above)
Comment (gnzlbg) [reply src/test/incremental/simd_intrinsic_ice.rs]: How tied is what this fixes to the simd intrinsics? E.g. if we refactor them, they might stop failing here, and a subsequent PR could reintroduce the hash collision.

Comment (oli-obk) [reply src/test/incremental/simd_intrinsic_ice.rs]: pretty much as long as you keep calling `const_field` it should ICE, and once you stop calling `const_field`, it stops needing to be a query and thus also can't ICE anymore

**Comment by oli-obk** [comment]
@bors retry 

> [01:59:42] error: linking with `cc` failed: signal: 4

File: src/librustc/mir/interpret/value.rs [hunk 3]
```diff
@@ -43,14 +43,21 @@ pub enum ConstValue<'tcx> {
...
 
-    /// An allocation together with a pointer into the allocation.
-    /// Invariant: the pointer's `AllocId` resolves to the allocation.
-    /// The alignment exists to allow `const_field` to have `ByRef` access to nonprimitive fields
-    /// of `repr(packed)` structs. The alignment may be lower than the type of this constant.
-    /// This permits reads with lower alignment than what the type would normally require.
-    /// FIXME(RalfJ,oli-obk): The alignment checks are part of miri, but const eval doesn't really
-    /// need them. Disabling them may be too hard though.
-    ByRef(Pointer, Align, &'tcx Allocation),
+    /// An value not represented/representable by `Scalar` or `Slice`
```
Comment (RalfJung) [inline src/librustc/mir/interpret/value.rs]:
```suggestion
    /// A value not represented/representable by `Scalar` or `Slice`
```

File: src/librustc/ich/impls_ty.rs [hunk 4]
```diff
@@ -175,13 +175,17 @@ impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::Allocation {
...
-        self.bytes.hash_stable(hcx, hasher);
-        for reloc in self.relocations.iter() {
+        let mir::interpret::Allocation {
+            bytes, relocations, undef_mask, align, mutability,
+            extra: _,
+        } = self;
+        bytes.hash_stable(hcx, hasher);
+        for reloc in relocations.iter() {
             reloc.hash_stable(hcx, hasher);
         }
```
Comment (RalfJung) [inline src/librustc/ich/impls_ty.rs]: Shouldn't this also hash the length? Otherwise, collisions might still be possible in theory.

**Pull request merged**